src/firstboot/firstboot.c

```diff
@@ -317,6 +318,166 @@ static int process_locale(void) {
+                for (;;) {
+                        _cleanup_free_ char *line = NULL;
+                        const char *l;
+
+                        r = read_line(original, LONG_LINE_MAX, &line);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                break;
+
+                        l = skip_leading_chars(line, "#");
+
+                        /* If this line isn't the locale we're looking for, write back the line
+                        unchanged. If it is the locale we're looking for, write back the line
+                        without leading whitespace and comment characters. */
+                        if (strncaseeq(locale, l, strlen(locale)))
+                                found = true;
+                        else
+                                l = line;
+
+                        fprintf(localegen, "%s\n", l);
+                }
+        else
+                if (errno != ENOENT)
+                        return -errno;
+
+        if (!found) {
+                const char *p, *charset;
+
+                /* Try to determine the charset from the given locale. */
+                p = strstr(locale, ".");
+                if (p)
+                        charset = p + 1;
+                else {
+                        log_info("Could not determine locale (%s) charset, assuming UTF-8", locale);
+                        charset = "UTF-8";
+                }
+
+                fprintf(localegen, "%s %s\n", locale, charset);
+        }
+
+        r = fflush_sync_and_check(localegen);
+        if (r < 0)
+                return r;
+
+        r = rename(localegen_tmp, localegen_path);
+        if (r < 0)
+                return -errno;
+
+        return 0;
+}
+
+static int invoke_locale_gen(void) {
+        int r;
+
+        r = safe_fork("(locale-gen)", FORK_RESET_SIGNALS|FORK_RLIMIT_NOFILE_SAFE|FORK_DEATHSIG|FORK_LOG|FORK_WAIT, NULL);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                if (arg_root) {
+                        r = chdir(arg_root);
+                        if (r < 0) {
+                                log_error_errno(errno, "Failed to change working directory to %s: %m", arg_root);
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        r = chroot(".");
+                        if (r < 0) {
+                                log_error_errno(errno, "Failed to chroot to %s: %m", arg_root);
+                                _exit(EXIT_FAILURE);
+                        }
+                }
+
+                execlp("locale-gen", "locale-gen", NULL);
+
+                log_error_errno(errno, "Failed to execute locale-gen: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        return 0;
+}
+
+static int process_locale_gen(void) {
+        const char *etc_localegen;
+        int r;
+
+        r = find_binary("locale-gen", NULL);
```

Comment:
I removed the `find_binary` check since I don't think we can call it after a fork since it allocates memory. I guess we can just fail if we don't find it? Unless we want to support doing the locale stuff on non-glibc systems. I'm not sure if that's supported.